dot/sync/chain_processor.go

```diff
@@ -108,101 +111,113 @@ func (s *chainProcessor) processReadyBlocks() {
 	}
-	hasBody, err := s.blockState.HasBlockBody(bd.Hash)
+
+	bodyInState, err := c.blockState.HasBlockBody(blockData.Hash)
 	if err != nil {
-		return fmt.Errorf("failed to check block state has body for hash %s: %w", bd.Hash, err)
+		return fmt.Errorf("checking if block state has body: %w", err)
 	}
 
 	// while in bootstrap mode we don't need to broadcast block announcements
-	announceImportedBlock := s.chainSync.syncState() == tip
-	if hasHeader && hasBody {
-		// TODO: fix this; sometimes when the node shuts down the "best block" isn't stored properly,
-		// so when the node restarts it has blocks higher than what it thinks is the best, causing it not to sync
-		// if we update the node to only store finalised blocks in the database, this should be fixed and the entire
-		// code block can be removed (#1784)
-		block, err := s.blockState.GetBlockByHash(bd.Hash)
+	announceImportedBlock := c.chainSync.syncState() == tip
+	if headerInState && bodyInState {
+		err = c.processBlockDataWithStateHeaderAndBody(blockData, announceImportedBlock)
 		if err != nil {
-			logger.Debugf("failed to get block header for hash %s: %s", bd.Hash, err)
-			return err
-		}
-
-		logger.Debugf(
-			"skipping block number %d with hash %s, already have",
-			block.Header.Number, bd.Hash) // TODO is this valid?
-
-		err = s.blockState.AddBlockToBlockTree(block)
-		if errors.Is(err, blocktree.ErrBlockExists) {
-			return nil
-		} else if err != nil {
-			logger.Warnf("failed to add block with hash %s to blocktree: %s", bd.Hash, err)
-			return err
+			return fmt.Errorf("processing block data with header and "+
+				"body in block state: %w", err)
 		}
+		return nil
+	}
 
-		if bd.Justification != nil {
-			logger.Debugf("handling Justification for block number %d with hash %s...", block.Header.Number, bd.Hash)
-			err = s.handleJustification(&block.Header, *bd.Justification)
-			if err != nil {
-				return fmt.Errorf("handling justification: %w", err)
-			}
+	if blockData.Header != nil && blockData.Body != nil {
+		err = c.processBlockDataWithHeaderAndBody(blockData, announceImportedBlock)
+		if err != nil {
+			return fmt.Errorf("processing block data with header and body: %w", err)
 		}
+		logger.Debugf("block with hash %s processed", blockData.Hash)
+	}
 
-		// TODO: this is probably unnecessary, since the state is already in the database
-		// however, this case shouldn't be hit often, since it's only hit if the node state
-		// is rewinded or if the node shuts down unexpectedly (#1784)
-		state, err := s.storageState.TrieState(&block.Header.StateRoot)
+	if blockData.Header != nil && blockData.Justification != nil && len(*blockData.Justification) > 0 {
+		err = c.handleJustification(blockData.Header, *blockData.Justification)
 		if err != nil {
-			logger.Warnf("failed to load state for block with hash %s: %s", block.Header.Hash(), err)
-			return err
+			return fmt.Errorf("handling justification: %w", err)
 		}
+	}
```

Comment:
Can we move to a single `if blockData.Header != nil {` block

Reply:
Indeed! Done :wink: